### PR TITLE
Remove ConfigurationBinder usage from Console Logging

### LINKED ISSUE #81931: Remove MS.Extensions.Logging.Console's dependency on ConfigurationBinder
In measuring the size of a NativeAOT published ASP.NET app, a considerable amount of size is added when using Console logging.

`dotnet new api -aot` and `dotnet publish` on win-x64: **13.2 MB**. Removing the line `builder.Logging.AddConsole()`: **12.1 MB**.

From my investigation, I've determined that ~860KB of size comes from using the ConfigurationBinder, which also pulls in Sytem.ComponentModel.TypeConverter. One reason this is so big is because a bunch of methods on `ICollection`/`IList` can be trimmed if we don't include ConfigurationBinder/TypeConverter. There are a lot of generic instantiations of collections (Dictionary, List, etc), and all these interface methods across scores of collection types add up. The methods in the `System.Collections.Generic` namespace drop by 343 KB alone.

For a full picture of all the code that can be trimmed here, diff the following mstat dumps:

[no-configbinder-console.txt](https://github.com/dotnet/runtime/files/10702618/no-config-console-collections.txt) was created using https://github.com/eerhardt/runtime/commit/af483745bed60cd2dcea797d5dda84de34c92a85.
[configbinder-console.txt](https://github.com/dotnet/runtime/files/10702620/console-collections.txt) was created from `main`

To eliminate this code from the application, we should remove the ConfigurationBinder usages in Logging.Console. They come from 2 places:

https://github.com/dotnet/runtime/blob/46cb4eddbb1f4bebdea970964d7449bacde8df1a/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.cs#L43

https://github.com/dotnet/runtime/blob/46cb4eddbb1f4bebdea970964d7449bacde8df1a/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerExtensions.cs#L141

And instead we can hand-write the "configuration => options" deserialization code. This can also improve startup performance, since we won't be using Reflection here.

Another option would be to use https://github.com/dotnet/runtime/issues/44493, but that isn't available yet. We can always hand-write the code for now, and use the source generator later once available.

### PR DESCRIPTION
**Pull request by eerhardt**
This allows ConfigurationBinder, and its dependencies like TypeConverter, to be trimmed in an application that uses Console Logging, like an ASP.NET API application.

Fix #81931

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerConfigureOptions.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            if (_configuration["UseUtcTimestamp"] is string useUtcTimestamp)
+            {
+#pragma warning disable CS0618 // Type or member is obsolete
+                options.UseUtcTimestamp = bool.Parse(useUtcTimestamp);
+#pragma warning restore CS0618 // Type or member is obsolete
+            }
+        }
+
+        public static T ParseEnum<T>(string value) where T : struct =>
+#if NETSTANDARD || NETFRAMEWORK
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerConfigureOptions.cs:90]:
```suggestion
#if NETFRAMEWORK || NETSTANDARD2_0
```

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            _configuration = providerConfiguration.GetFormatterOptionsSection();
+        }
+
+        public void Configure(ConsoleFormatterOptions options) => Bind(_configuration, options);
+
+        public static void Bind(IConfiguration configuration, ConsoleFormatterOptions options)
+        {
+            if (configuration["IncludeScopes"] is string includeScopes)
+            {
+                options.IncludeScopes = bool.Parse(includeScopes);
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs:35]: We have good tests in place for things like what happens if parsing fails?

Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs:35]: In configuration binding the TypeConverters are used and can throw which I believe we wrap the exception inside InvalidOperationException.

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerConfigureOptions.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerConfigureOptions.cs:90]: Isn't `NETFRAMEWORK` will included with  `NETSTANDARD2_0`? or we have check for both?

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        private readonly IConfiguration _configuration;
+
+        public ConsoleFormatterConfigureOptions(ILoggerProviderConfiguration<ConsoleLoggerProvider> providerConfiguration)
+        {
+            _configuration = providerConfiguration.GetFormatterOptionsSection();
+        }
+
+        public void Configure(ConsoleFormatterOptions options) => Bind(_configuration, options);
+
+        public static void Bind(IConfiguration configuration, ConsoleFormatterOptions options)
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs:31]: >Bind [](http://example.com/codeflow?start=27&length=4)

Why defining Bind method and not implemented this inside the Configure method?

Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs:31]: never mind, I am seeing you are calling it from the json formatter one.

File: src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerExtensionsTests.cs [hunk 4]
```diff
@@ -374,8 +374,7 @@ public void AddConsole_MaxQueueLengthSetToNegativeOrZero_Throws(int invalidMaxQu
                 )
                 .BuildServiceProvider();
 
-            // the configuration binder throws TargetInvocationException when setting options property MaxQueueLength throws exception
-            Assert.Throws<TargetInvocationException>(() => serviceProvider.GetRequiredService<ILoggerProvider>());
+            Assert.Throws<ArgumentOutOfRangeException>(() => serviceProvider.GetRequiredService<ILoggerProvider>());
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerExtensionsTests.cs:377]: >ArgumentOutOfRangeException [](http://example.com/codeflow?start=26&length=27)

should we keep the old exception in such case?

**Comment by tarekgh** [comment]
Changes LGTM. do we have tests when having invalid values inside the configuration? something when defining string with `null` value (which we convert it to empty string. or when setting `bool` value to something different than `true` or `false`. Or having wrong int values. I am asking to ensure the behavior does not change when encountering such cases.

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs:35]: > We have good tests in place for things like what happens if parsing fails?

No, but I can add some tests.

> which I believe we wrap the exception inside InvalidOperationException.

I don't think we should strive to maintain the same exception types when the configuration is mis-configured in this scenario. The configuration is bound lazily, and so the exception doesn't occur until someone actually tries getting an `ILogger` service. It is not predictable when the exception will be thrown, so it wouldn't be possible for someone to try to `catch` the InvalidOperationException. The important piece is that an actionable/understandable exception is thrown.

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerConfigureOptions.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerConfigureOptions.cs:90]: You need to check for both. We build for the following TFMs:

https://github.com/dotnet/runtime/blob/8ad7113ac3e27453d3b9664181c0a5297323ae65/src/libraries/Microsoft.Extensions.Logging.Console/src/Microsoft.Extensions.Logging.Console.csproj#L4

File: src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerExtensionsTests.cs (hunk 4 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerExtensionsTests.cs:377]: `TargetInvocationException` shouldn't have been thrown in the first place. Maybe that is a bug we should log on ConfigurationBinder. We should be passing `BindingFlags.DoNotWrapExceptions`.

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs:35]: The one piece where the existing exception is useful is that passes in the configuration `path` to the exception message. So you get an exception like:

```
System.InvalidOperationException: Failed to convert configuration value at 'FormatterOptions:SingleLine' to type 'System.Boolean'.
 ---> System.FormatException: not a bool is not a valid value for Boolean.
 ---> System.FormatException: String 'not a bool' was not recognized as a valid Boolean.
   at System.Boolean.Parse(ReadOnlySpan`1 value)
   at System.Boolean.Parse(String value)
```

`FormatterOptions:SingleLine` is the key piece of info we would be missing. Versus the current PR behavior:

```
Unhandled exception. System.FormatException: String 'not a bool' was not recognized as a valid Boolean.
   at System.Boolean.Parse(ReadOnlySpan`1 value)
   at System.Boolean.Parse(String value)
   at Microsoft.Extensions.Logging.SimpleConsoleFormatterConfigureOptions.Configure(SimpleConsoleFormatterOptions options) in C:\git\runtime\src\libraries\Microsoft.Extensions.Logging.Console\src\SimpleConsoleFormatterConfigureOptions.cs:line 40
```

**Comment by layomia** [comment]
Overall LGTM.

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerConfigureOptions.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerConfigureOptions.cs:90]: Yes, I don't know if the SDK is smart enough in defining these. I mean if defining `NETFRAMEWORK` is covering `NETSTANDARD2_0`. But anyway checking both never hurt :-)

File: src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerExtensionsTests.cs (hunk 4 above)
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Console/tests/Microsoft.Extensions.Logging.Console.Tests/ConsoleLoggerExtensionsTests.cs:377]: I was only asking if we should be concerned about app compatibility. I think it is very unlikely to have anyone depend on such thrown exception though.

**Review by tarekgh — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterConfigureOptions.cs]: I've addressed this with the latest commit. PTAL.

**Comment by tarekgh** [comment]
LGTM!

**Comment by ericstj** [comment]
@eerhardt would we just replace this with the Configuration Binder source generator generated code once that's ready?

**Comment by eerhardt** [comment]
> @eerhardt would we just replace this with the Configuration Binder source generator generated code once that's ready?

Yes, that is the goal/intention. See https://github.com/dotnet/runtime/issues/81931#issue-1578811453. If/when the Configuration Binder source generator can support this library, we should switch to using it.

Things that may be tricky:

* `netstandard2.0` and netfx support
* Properties that are annotated with ObsoleteAttribute still need to be bound
* The mutable struct `JsonWriterOptions` needs to be supported

**Pull request merged**